### PR TITLE
Fix deadlock on delegate target subscription

### PR DESCRIPTION
**Pull request by tamasseller**
Fix deadlock when the target of the delegate tries to access the WeakEventSource by breaking up the big lock block into two shorter. Not tested thoroughly, but should be safe, because only locals are accessed outside of the two lock blocks.

### CODE REVIEW

**Comment by thomaslevesque** [comment]
Hi @tamasseller, thanks for the PR!

Could you please show an example where the deadlock occurs?

**Comment by tamasseller** [comment]
Hi,

I attached a stripped down version of what i was trying to acheive. It hangs with the original version, and works fine with my modification. It may seem rather arbitrary but the business constraints force this structure on me.

Briefly: there is a worker thread that receives work items, some of which are required to complete before the sender of the item can continue. If thread A executes a callback through Raise that posts a work item for thread B and waits for its completion and than the work on thread B tries to access the WeakEventSource - for example to Subscribe to it - then they both try to lock on the _handlers object and deadlock occurs.

[Program.zip](https://github.com/thomaslevesque/WeakEvent/files/2542286/Program.zip)

**Comment by thomaslevesque** [comment]
Hi @tamasseller,

Thanks for the repro. The example is a bit hard to follow, but I see what the problem is. I agree that keeping the lock while the handlers are invoked is bound to cause problems. However your fix can cause incorrect behavior: you find the indices of failed handlers on a copy of the `_handlers`, but since you don't have the lock, some other thread could modify `_handlers` in the meantime, so the indices in `failedHandlers` might no longer be correct in the loop that calls `Invalidate`.

So I can't merge your PR as is, but I'll try to implement a fix.

**Comment by thomaslevesque** [comment]
Hi @tamasseller,

I just released 2.3.0-beta001 which should fix the problem (at least it did with the repro you provided).
Could you please check if it works for you?
